### ISSUE TITLE
Keep previous FSRS parameters if they get worse when optimizing

### ISSUE THREAD

**Issue by dae** [ankitects/anki#2972]
https://forums.ankiweb.net/t/measures-to-prevent-new-fsrs-parameters-from-being-worse-than-the-old-ones/38957

**Comment by Expertium** [comment]
Previously, you said, "Before we can implement this, we’ll need to know how we should handle RMSE/log-loss moving in different directions." I think the right thing is "Only look at RMSE." The problem with log-loss is that it depends on retention. RMSE (bins), on the other hand, does not depend on retention.

**Comment by dae** [comment]
Possible approach we could use:

- have compute_weights take existing weights as an argument (with defaults being used if blank)
- calculate RMSE prior to computing new weights
- calculate RMSE for new weights, and return an error if they're worse (perhaps allowing up to a relative 5% regression)

**Comment by Expertium** [comment]
> * and return an error if they're worse

That would just confuse the user. I think instead it's better to just use automatically "Evaluate" with both old and new parameters, and then keep whichever are better. All of that should happen without notifying the user. Even if we notified the user, it wouldn't really help them to make any kind of good decision, it would just make them wonder "Huh? Is the optimizer broken? Is there a problem with my settings?".
@L-M-Sherlock what do you think?

**Comment by dae** [comment]
Sorry, I should have been more explicit. I was expecting the frontend to catch the returned error, and skip updating the existing weights in that case.

**Comment by Expertium** [comment]
By the way, according to [this comment of LMSherlock](https://github.com/open-spaced-repetition/fsrs4anki/issues/567#issuecomment-1872866452), it doesn't matter whether the default parameters or the previous ones are used when initializing the optimizer.

**Comment by L-M-Sherlock** [comment]
> The problem with log-loss is that it depends on retention

It doesn't matter in this case, because we evaluate the parameters in the same dataset.

**Comment by user1823** [comment]
@L-M-Sherlock, do you think that we should only look at the RMSE (bins)? If so, why?

I want to add that I re-optimized my parameters today and got the following result:
- Log loss went from 0.2047 to 0.2042
- RMSE (bins) went from 0.84% to 1.02%

In this case, the relative change in RMSE is larger. So, I think that the parameters are worse overall.

But, if the relative change in log loss is larger in magnitude, should we still consider only RMSE to decide whether to update the parameters or not?

**Comment by L-M-Sherlock** [comment]
According to my previous experiment, the RMSE(bins) can be decreased even when the parameters are changed very slightly:
https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/blob/main/ICILoss.ipynb

**Comment by Expertium** [comment]
> It doesn't matter in this case, because we evaluate the parameters in the same dataset.

It does. If the user's retention consistently goes down, log-loss will keep getting worse after each optimization, bur RMSE (bins) won't necessarily keep getting worse, or at least not as much.

**Comment by L-M-Sherlock** [comment]
> It does. If the user's retention consistently goes down, log-loss will keep getting worse after each optimization, bur RMSE (bins) won't necessarily keep getting worse, or at least not as much.

The log-loss of old parameters is also calculated in the current dataset whatever its retention goes up or down.

**Comment by Expertium** [comment]
I don't understand what you're trying to say.

**Comment by user1823** [comment]
No reviews are done between the two evaluations. So, the retention must be the same.

**Comment by Expertium** [comment]
Ah. Yeah, that's true. I was talking about a more complicated scenario, where the parameters are re-optimized many times and each time the new log-loss is greater. But it's possible that the new RMSE wouldn't be greater each time.

**Comment by user1823** [comment]
Well, I didn't get what you are saying. But, you don't need to explain. (Chances are high that I still won't be able to draw any conclusions from that)

Can there be any case where the log loss decreases and the RMSE increases but the parameters are more optimal? If not, we can simply use RMSE as the only criteria to decide whether the new parameters are better or worse.

**Comment by L-M-Sherlock** [comment]
> Can there be any case where the log loss decreases and the RMSE increases but the parameters are more optimal?

Theoretically, it exists: https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/blob/main/metrics_research.ipynb

**Comment by user1823** [comment]
Can you point out which one? Just edit your previous comment to include the example instead of creating a new comment. I don't want to pollute the comments list.

**Comment by user1823** [comment]
It seems that we can rely solely on RMSE(bins).

The case of an algorithm giving the same prediction for all the reviews doesn't seem to be practical. Even if there was such a case where a very large number of reviews were put in the same bin, it seems to be difficult to detect.

Of course, we can change the criteria later if we have more insights. But for now, I think that only using RMSE(bins) is the way to go.

**Comment by user1823** [comment]
> All of that should happen without notifying the user.

I think that we can show a message similar to "Your parameters are already optimal." This way,
- they would not be left wondering why the parameters didn't change; 
- they would get to know that re-optimizing that frequently is not required.

**Comment by Expertium** [comment]
> Well, I didn't get what you are saying. But, you don't need to explain. (Chances are high that I still won't be able to draw any conclusions from that)

I meant a situation like this. As n reviews increases:
1) Retention decreases
2) Log-loss increases
3) RMSE (bins) stays the same

![image](https://github.com/ankitects/anki/assets/83031600/86634707-0cb6-4bae-b534-282546dab6f8)
Here's a toy example. The numbers aren't shown because the exact values don't matter, what matter is the slope of the curve. Retention goes down, log-loss goes up, RMSE (bins) stays the same. This seems like a plausible scenario to me, which is why I recommend focusing only on RMSE (bins).

**Comment by user1823** [comment]
> As n reviews increases:

But in the current problem, n_reviews remains the same but the parameters change. (The two evaluations run on exactly the same data, just with different parameters)

So, what you have described doesn't seem to be applicable in the current situation.

**Comment by Expertium** [comment]
> I think that we can show a message similar to "Your parameters are already optimal."

I like that idea. So @L-M-Sherlock, me and user1823 agree that only RMSE (and not log-loss) should be used to decide which parameters to keep, so if you don't have any objections, you should probably start working on implementing it.
Also, about this:
![image](https://github.com/ankitects/anki/assets/83031600/7e0cabf3-63bb-4f98-8635-835ae02d771a)
In the latter case, not only RMSE but every other metric is also worse. R-squared is worse, ICI-related metrics are all worse, and the slope and intercept are further from 1 and 0 (which corresponds to a perfect algorithm). So I would say that the latter parameters are worse.

**Comment by L-M-Sherlock** [comment]
> So I would say that the latter parameters are worse.

But the weird thing is the first one's parameters only change sightly:

<img width="917" alt="image" src="https://github.com/ankitects/anki/assets/32575846/1e9ace18-d4c2-4c07-bb99-27cce0232d6f">

It's w[16] is still near 2.78.

<img width="917" alt="image" src="https://github.com/ankitects/anki/assets/32575846/64a740f1-9a64-4834-aa22-d51e1dacbf3e">

The second one's w[16] reach its upper limit rapidly.

**Comment by L-M-Sherlock** [comment]
Fine. I don't have any objections.

**Comment by user1823** [comment]
SuperMemo says that the universal metric is the best way to compare two spaced-repetition algorithms.

@L-M-Sherlock, can the UM be calculated as quickly as calculating RMSE twice? If so, we can use UM to compare the old and the new weights.

If not, RMSE(bins) would be fine.

**Comment by L-M-Sherlock** [comment]
> can the UM be calculated as quickly as calculating RMSE twice

Of course. I have implemented it in the Python Optimizer. I plan to port it into fsrs-rs.

**Comment by dae** [comment]
When implementing this, we'll want to update to fsrs 0.3: https://github.com/open-spaced-repetition/fsrs-rs/pull/154

**Comment by Expertium** [comment]
> The new parameters have worse RMSE than the RMSE of the previous parameters when they were generated.

No, both sets of parameters are evaluated on the full dataset. In your example, both sets of parameters would be evaluated on the 10000 reviews dataset. We do not compare RMSE(old parameters, 1 review) to RMSE(new parameters, 10000 reviews). We compare RMSE(old parameters, 10000 reviews) to RMSE(new parameters, 10000 reviews).

**Comment by Xemorr** [comment]
> > The new parameters have worse RMSE than the RMSE of the previous parameters when they were generated.
> 
> No, both sets of parameters are evaluated on the full dataset. In your example, both sets of parameters would be evaluated on the 10000 reviews dataset. We do not compare RMSE(old parameters, 1 review) to RMSE(new parameters, 10000 reviews). We compare RMSE(old parameters, 10000 reviews) to RMSE(new parameters, 10000 reviews).

Thank you for the clarification, I was having a conversation with that prolific r/anki redditor and he didn't have a proper answer for me. So this means it's the case where the optimiser does not always give the optimal parameters for the training data. Why is that (is there a training/validation split)?

**Comment by Expertium** [comment]
> I was having a conversation with that prolific r/anki redditor

u/ClarityInMadness? That's me. Sorry if I missed your comment.

> So this means it's the case where the optimiser does not always give the optimal parameters for the training data. Why is that?

Honestly, ask @L-M-Sherlock. I don't fully understand why myself.

**Comment by Xemorr** [comment]
> > I was having a conversation with that prolific r/anki redditor
> 
> u/ClarityInMadness? That's me. Sorry if I missed your comment.
> 
> > So this means it's the case where the optimiser does not always give the optimal parameters for the training data. Why is that?
> 
> Honestly, ask @L-M-Sherlock. I don't fully understand why myself.

oh this was a while ago, I don't think you knew why at the time.

**Comment by L-M-Sherlock** [comment]
> So this means it's the case where the optimiser does not always give the optimal parameters for the training data. Why is that

Gradient descent-based methods are not guaranteed to find the global optimal  parameters.

**Comment by user1823** [comment]
@L-M-Sherlock, I agree that this hack was needed to deal with the issue that you just mentioned.

However, there is a high probability that the latest version has **another issue** that makes it more likely to get worse parameters after optimization. For instance, I have never been able to get better parameters than those I got in the end of December (when 23.12.1 was released).

In my opinion, the current issue is not the right place to discuss this further. So, if you want more details and/or want to discuss something, I recommend creating a new issue in the FSRS repo.

**Comment by Xemorr** [comment]
> > So this means it's the case where the optimiser does not always give the optimal parameters for the training data. Why is that
> 
> Gradient descent-based methods are not guaranteed to find the global optimal  parameters.

Thanks Jarrett, I wasn't sure what kind of optimisation was being done. That makes sense
